### PR TITLE
Stopgap: PostgreSQL constraint deferring

### PR DESCRIPTION
**Pull request by pimeys**
Part of: https://github.com/prisma/schema-team/issues/386

### CODE REVIEW

File: schema-engine/connectors/sql-schema-connector/src/warnings/generators.rs [hunk 1]
```diff
@@ -275,6 +279,15 @@ pub(crate) struct ModelAndIndex {
     pub(crate) index_db_name: String,
 }
 
+/// An constraint in a model that triggered a warning.
```
Comment (Druue) [inline schema-engine/connectors/sql-schema-connector/src/warnings/generators.rs:282]:
```suggestion
/// A constraint in a model that triggered a warning.
```

File: schema-engine/connectors/sql-schema-connector/src/warnings/generators.rs [hunk 2]
```diff
@@ -275,6 +279,15 @@ pub(crate) struct ModelAndIndex {
...
+#[derive(Serialize, Debug, Clone)]
+pub(crate) struct ModelAndConstraint {
+    /// The name of the model
+    pub(crate) model: String,
+    /// The name of the index
+    pub(crate) constraint: String,
+}
```
Comment (Druue) [inline schema-engine/connectors/sql-schema-connector/src/warnings/generators.rs:289]: I'm a little confused here as the struct is `ModelAndConstraint` and the field is `constraint` but the doc is saying name of the _index_. These are not necessarily the same thing right

File: schema-engine/sql-schema-describer/src/postgres/indexes_query.sql [hunk 3]
```diff
@@ -42,5 +44,6 @@ FROM
     INNER JOIN pg_am AS indexaccess ON indexaccess.oid = indexinfo.relam
     LEFT JOIN pg_opclass AS opclass -- left join because crdb has no opclasses
         ON opclass.oid = rawindex.indclass
+    LEFT JOIN pg_constraint pc ON rawindex.indexrelid = pc.conindid AND pc.contype <> 'f'
```
Comment (Druue) [inline schema-engine/sql-schema-describer/src/postgres/indexes_query.sql:47]: Do we have anywhere / could you point me in the right direction for where I can read up about these?
There just being an exclusion check for `f` seems very arcane and magical to me 🤔

File: schema-engine/connectors/sql-schema-connector/src/warnings/generators.rs (hunk 2 above)
Comment (pimeys) [reply schema-engine/connectors/sql-schema-connector/src/warnings/generators.rs:289]: Copy&pasta. It should say constraint.

File: schema-engine/connectors/sql-schema-connector/src/warnings/generators.rs [hunk 4]
```diff
@@ -275,6 +279,15 @@ pub(crate) struct ModelAndIndex {
     pub(crate) index_db_name: String,
 }
 
+/// An constraint in a model that triggered a warning.
+#[derive(Serialize, Debug, Clone)]
+pub(crate) struct ModelAndConstraint {
+    /// The name of the model
+    pub(crate) model: String,
+    /// The name of the index
```
Comment (pimeys) [inline schema-engine/connectors/sql-schema-connector/src/warnings/generators.rs:287]:
```suggestion
    /// The name of the constraint
```

File: schema-engine/sql-schema-describer/src/postgres/indexes_query.sql (hunk 3 above)
Comment (pimeys) [reply schema-engine/sql-schema-describer/src/postgres/indexes_query.sql:47]: Hehe, yes it is arcane magic. So for foreign keys we get an index constraint to the table. If we don't filter them out, we get some indices twice :P

Comment (Druue) [reply schema-engine/sql-schema-describer/src/postgres/indexes_query.sql:47]: ahh, so f == is the foreign key constraint type
I'm guessing that's what `.contype` means

**Review by Druue — APPROVED** [review]

File: schema-engine/sql-schema-describer/src/postgres/indexes_query.sql (hunk 3 above)
Comment (janpio) [reply schema-engine/sql-schema-describer/src/postgres/indexes_query.sql:47]: I think you are looking for this @Druue : https://www.postgresql.org/docs/current/catalog-pg-constraint.html

Comment (pimeys) [reply schema-engine/sql-schema-describer/src/postgres/indexes_query.sql:47]: Yeah, and we don't care about those in the index query.

**Pull request merged**